lib/src/repo.rs

```diff
@@ -242,38 +245,30 @@ impl ReadonlyRepo {
         })
     }
 
-    // An interface for testing this functionality directly is constructed in
-    // a follow-up commit.
+    fn commit_change_id_index(&self) -> &IdIndex {
```

Comment:
Since the BTree version is both simpler and better than the naive version, I'd just squash it into that version. But I would still prefer that it's abstracted in the Index type.

By the way, we'll also want a higher-level function (outside the index) for first attempting to resolve a prefix within an configured revset (perhaps `ui.default-revset`). I don't think that will impact the interface we need from the index, however - we'll probably just want exactly the signature of your current `shortest_unique_prefix_length()`.

Reply:
> By the way, we'll also want a higher-level function (outside the index) for first attempting
> to resolve a prefix within an configured revset (perhaps ui.default-revset)

It's important that whatever does this is used both for finding the shortest possible prefix of a commit and for looking up commit by its prefix.

I'm not sure if revsets are the best for this -- they don't currently cover invisible commits. I was thinking along the lines of restricting to all of your commits for the past year and all of their non-public ancestors/descendants, and all entries in the obslog for these. But all of this can be decided later.

I originally thought we'd have a sequence of larger and larger tries to do this. I think you have a different idea in mind.

Reply:
> It's important that whatever does this is used both for finding the shortest possible prefix of a commit and for looking up commit by its prefix.

True. The lookup would probably have to be somewhere around here: https://github.com/martinvonz/jj/blob/24ccd80a5c0b3c4ad7306b7db35adf28a2747a91/lib/src/revset.rs#L193-L203

> I'm not sure whether revsets are the best for this -- they don't currently cover invisible commits, and I'd like to have easy access to at least recent entries in my obslog (though I wouldn't say it's critical, if this desire makes life harder).

We could add revsets that access the operation log in order to find those. At Google, our default revset is small enough that the unique prefix is one or two hex digits for most users' workflows.

Reply:
> Since the BTree version is both simpler and better than the naive version, I'd just squash it into that version

I don't know if that's worth doing at this point. Let me know if you'd really like me to.

Reply:
Nah, that's fine. I've already reviewed both versions now, but it would have been better to not have to review the naive version :)

Reply:
Sorry about that. I'll try to make some future PR quick and efficient to review instead of planning on it being a huge pain.

Reply:
No problem! It really wasn't a huge pain to review this PR :)

Reply:
I'm quite happy to hear that!